core/src/loader.rs

```diff
@@ -2107,4 +2217,449 @@ impl<'gc> Loader<'gc> {
+                                                &[target_object.into()],
+                                                "onIOError".into(),
+                                            )?;
+                                        }
+                                        _ => {
+                                            tracing::warn!(
+                                                "Unhandled non-fetch error on download: {:?}",
+                                                err.error
+                                            );
+                                        }
+                                    }
+                                }
+                            }
+                        } else {
+                            as_broadcaster::broadcast_internal(
+                                &mut activation,
+                                target_object,
+                                &[target_object.into()],
+                                "onCancel".into(),
+                            )?;
+                        }
+                    }
+                    Err(err) => {
+                        tracing::warn!("Download dialog had an error {:?}", err);
+                    }
+                }
+
+                Ok(())
+            })
+        })
+    }
+
+    /// Loader to handle a file upload task
+    ///
+    /// Uploads the given `data` to the provided `url`.
+    /// `file_name` is sent along with the data, as part of the multipart/form-data body
+    pub fn file_upload_loader(
+        &mut self,
+        player: Weak<Mutex<Player>>,
+        url: String,
+        data: Vec<u8>,
+        file_name: String,
+    ) -> OwnedFuture<(), Error> {
+        let handle = match self {
+            Loader::UploadFile { self_handle, .. } => {
+                self_handle.expect("Loader not self-introduced")
+            }
+            _ => return Box::pin(async { Err(Error::NotFileUploadLoader) }),
+        };
+
+        let player = player
+            .upgrade()
+            .expect("Could not upgrade weak reference to player");
+
+        Box::pin(async move {
+            let total_size_bytes = data.len();
+
+            // Format the data into multipart/form-data
+            let mut out_data = Vec::new();
+            out_data.extend_from_slice(b"------------BOUNDARY\n");
+            out_data.extend_from_slice(b"Content-Disposition: form-data; name=\"Filename\"\n\n");
+            out_data.extend_from_slice(file_name.as_bytes());
+            out_data.extend_from_slice(b"\n------------BOUNDARY\n");
+            out_data.extend_from_slice(
+                b"Content-Disposition: form-data; name=\"Filedata\"; filename=\"",
+            );
+            out_data.extend_from_slice(file_name.as_bytes());
+            out_data.extend_from_slice(b"\"\n");
+            out_data.extend_from_slice(b"Content-Type: application/octet-stream\n\n");
+            out_data.extend_from_slice(&data);
```

Comment:
What if the data contains "`------------BOUNDARY`"?

Reply:
Currently, that will break the form. One of the issues with this being such an old PR is that I don't remember if that is the value Flash uses or if It's invented. I guess this needs more testing

Reply:
I would just add a FIXME notice and leave it as so, or abort with an error if the data contains `------------BOUNDARY`.